### DISCUSSION TITLE
Make ConfigWatcher simpler

### DISCUSSION THREAD

**Discussion by pg-yang** [apache/skywalking#9670] [Ideas]
Hi , team 
Currently , if we want to add a dynamic config for OAP , we need to create class which extends `ConfigChangeWatcher`, and add it to  ModuleConfig  as field , just  like:
```java 
    @Setter
    @Getter
    private String slowCacheWriteThreshold = "default:20,redis:10";

    @Setter
    @Getter
    private CacheWriteLatencyThresholdsAndWatcher cacheWriteLatencyThresholdsAndWatcher;
```
And we have to set `cacheWriteLatencyThresholdsAndWatcher` on module prepare stage ,  register this watcher  through  `registerConfigChangeWatcher` on module start stage . 

This is initial idea about simplifying watcher in module config .
Declare interface in `library-module`, this interface is responsible for transforming property value to an object : 
```java 
public interface ModuleConfigurationProcessor {
    /**
     * @param propertyValue the value parsed from config property
     * @param targetField   the target of the value would inject
     * @return The final value would be injected to targetField
     */
    Object process(Object propertyValue, Field targetField);
}
```
`ModuleConfigurationProcessor`  should be defined in `ModuleDefine`  as field , and would be invoked  in method `ModuleDefine#copyProperties`
```java 
    private void copyProperties(ModuleConfig dest, Properties src, String moduleName,
                                String providerName) throws IllegalAccessException {
                //....
                Field field = getDeclaredField(destClass, propertyName);
                field.setAccessible(true);
                if (this.moduleConfigurationProcessor != null) {
                    Object propertyValue = this.moduleConfigurationProcessor.process(src.get(propertyName), field);
                    field.set(dest, propertyValue);
                }else{
                    field.set(dest, src.get(propertyName));
                }
              //......
        }
    }
```

Provide a annotation  , such as  `WatcherBinder`   in `configuration-api` module . The annotation could marked on field which type  is `ConfigChangeWatcher` subclass 
```java 
    @Setter
    @Getter
    @WatcherBinder
    private CacheWriteLatencyThresholdsAndWatcher slowCacheWriteThreshold;
```
Provide a implementation of `ModuleConfigurationProcessor`  in `configuration-api` module   and set the processor to  every 
 `ModuleDefine` when config module start , 
In the end , `ModuleConfigurationProcessor` could  transform  propertyValue  to a Watcher object  according  annotation `WatcherBinder` marked on the  targetField , also collect these watchers for registering

**Comment by wu-sheng** [comment]
I can get your idea. Basically, I think it should be good. But there are questions.
Is `@WatcherBinder` and `ModuleConfigurationProcessor` duplicated? As this field has to implement one kernel-level interface, which provides the module manager the possibility to initialize this.

**Comment by pg-yang** [comment]
Yes , `@WatcherBinder` and  `ModuleConfigurationProcessor ` may be duplicated . In fact , we could transform property value  according to the type of field .

**Reply by wu-sheng** [reply]
For the interface, I think it should be named `ConfigValueHolder`.  And I change your interface a little. Also, the property value could be `string` for less change to face converting error.

```java
/**
  * ConfigValueHolder implementation could be declared in the {@link ConfigValueHolder) as the type of a field.
  * The module manager would create this implementation object through default constructor 
  * and inject the initial value before assigning the object to the target field.
  */
public interface ConfigValueHolder {
    /**
     * Initial this holder object through the value from configuration file 
     *
     * @param propertyValue the value parsed from config property
     * @param targetField represents a field in the ConfigValueHolder implementing this interface.
     * @return the initialized field value to the given target field
     */
    Object init(String propertyValue, Field targetField);
}
```

WDYT?

**Reply by pg-yang** [reply]
Do you mean dynamic filed have to implement `ConfigValueHolder ` ?  The config filed needn't to extend any interface/supper class  unless  `ConfigChangeWatcher`.

The more rigorous interface  I wanted 

```java

public interface ModuleConfigurationProcessor {
    /**
     * @param propertyValue the value parsed from config property
     * @param targetField   the target of the value would inject
     * @return The final value would be injected to targetField
     */
    T <T extend ConfigChangeWatcher> process(Object propertyValue, Field targetField);
}
```

In fact , I create  `ModuleConfigurationProcessor ` for decoupling , because  `parsing property ` stage exist in `library-module`, but `ConfigWatcher` exist in  `configuration-api` .  `ModuleConfigurationProcessor  is defined in `library-module` . 
`configuration-api` could set WatcherProcessor (a  implementation of  ModuleConfigurationProcessor) to `library-module`(e. g. ModuleManager instance ) when prepare .

```java
public class ConfigurationModule extends ModuleDefine {
    public static final String NAME = "configuration";
    private ModuleConfigurationProcessor processor = new WatcherProcessor()

    public ConfigurationModule() {
        super(NAME);
    }

    void prepare(ModuleManager moduleManager, ApplicationConfiguration.ModuleConfiguration configuration,
                 ServiceLoader<ModuleProvider> moduleProviderLoader) throws ProviderNotFoundException, ServiceNotProvidedException, ModuleConfigException, ModuleStartException {
           supper. prepare();
         List< ModuleDefine> modules =  moduleManager.findAllModule();
        modules.foreach(module ->setModuleConfigurationProcessor(this.processor));     

}
    @Override
    public Class[] services() {
        return new Class[] {DynamicConfigurationService.class};
    }

}
```
ModuleDefine would update : 

```java 

public abstract class ModuleDefine implements ModuleProviderHolder {
     // omit setter 
    private ModuleConfigurationProcessor moduleConfigurationProcessor =null;
    
    private void copyProperties(ModuleConfig dest, Properties src, String moduleName,
                                String providerName) throws IllegalAccessException {
        if (dest == null) {
            return;
        }
        Enumeration<?> propertyNames = src.propertyNames();
        while (propertyNames.hasMoreElements()) {
            String propertyName = (String) propertyNames.nextElement();
            Class<? extends ModuleConfig> destClass = dest.getClass();
            try {
                Field field = getDeclaredField(destClass, propertyName);
                field.setAccessible(true);
                if (moduleConfigurationProcessor != null) {
                    Object propertyValue = moduleConfigurationProcessor.process(src.get(propertyName), field);
                    field.set(dest, propertyValue);
                }else{
                    field.set(dest, src.get(propertyName));
                }
            } catch (NoSuchFieldException e) {
                LOGGER.warn(
                    propertyName + " setting is not supported in " + providerName + " provider of " + moduleName + " module");
            }
        }
    }

}

```

The dynamic field would declare as 

```
  // Also   , we  provide feature for parsing string to an object (e,g, ParsedData)
  private StandardConfigChangeWatcher<ParsedData> cacheWriteLatencyThresholdsAndWatcher;
```

StandardConfigChangeWatcher  declared like 

```java 
class StandardConfigChangeWatcher<T> extends ConfigChangeWatcher{
   public T  getValue (){
      return value ;
}

}
```

In the end , all  module initialized behind  `configuration-api` would be aware of `moduleConfigurationProcessor`

**Reply by wu-sheng** [reply]
> Do you mean dynamic filed have to implement ConfigValueHolder ? 

No, I am not. This doesn't have to be obligate. I just polished your interface.

**Reply by pg-yang** [reply]
OK, it's unimportant . Let's talk about this later (or  in coding stage ).
How about the code structure ?

**Reply by wu-sheng** [reply]
>  T <T extend ConfigChangeWatcher> process(Object propertyValue, Field targetField);

I don't like this, as the dynamic configuration is not a part of the module management mechanism at all.

I think the original proposal is good enough, which is why I just polished your interface with more comments and some renaming.
We don't need to change many things as this proposal could land easily and no big impact.

**Reply by pg-yang** [reply]
Could you explain  the meaning of  `ConfigValueHolder ` you wanted ? I don't  understand it .

**Reply by wu-sheng** [reply]
`ConfigValueHolder` is actually the interface of your `ModuleConfigurationProcessor`. Nothing changed, it is just a rename.

**Reply by pg-yang** [reply]
> I don't like this, as the dynamic configuration is not a part of the module management mechanism at all.

Yes  , you are right , the `ConfigChangeWatcher ` should not be here .

And  is there other problems about the  above I  described in this Thread ?

**Reply by wu-sheng** [reply]
I am not sure the point of this

> modules.foreach(module ->setModuleConfigurationProcessor(this.processor));   

In the `#copyProperties`, you could detect the field type `instanceof` `ConfigValueHolder` interface directly, and initialize it through your proposed way. I am not sure why we need more `ModuleConfigurationProcessor` mechanism.

**Reply by pg-yang** [reply]
In fact , my  goal is converting  dynamic property value to a Watcher object  , like 
```java 
 private StandardConfigChangeWatcher<ParsedData> cacheWriteLatencyThresholdsAndWatcher;
```
But  I couldn't  write Watcher to  ` Module Manager`  module , because of  module  partitioning .
So I provide `ModuleConfigurationProcessor ` in   ` Module Manager`  module , and  `configuration-api` module could set an  implementation of `ModuleConfigurationProcessor`

**Reply by wu-sheng** [reply]
I think I may move too fast. You could have a `ConfigValueHolder` in `library-module`, and when initializing field type, if the type `instanceof`s the `ConfigValueHolder` interface, you use `class#newInstance` and call `ConfigValueHolder#init` to get an instance.

Then the watcher in the module/provider, you could implement watcher and ConfigValueHolder at the same time, and build a real watcher in the `init` method.

Are you following this now? `StorageDataComplexObject` interface is being used in a similar way in OAP storage layer.

**Reply by pg-yang** [reply]
I could understand a little . 
I put `WatcherConfigValueHolder ` to `configuration-api` as sub-interface  of `ConfigValueHolder `
```java
public interface WatcherConfigValueHolder extends ConfigValueHolder {
    @Override
    default Object init( Object propertyValue, Field field) {
      // ... 
     return new Watcher();
    }
}
```

all ModuleConfig which implement  `WatcherConfigValueHolder`  would be injected appropriate Watcher

**Reply by wu-sheng** [reply]
WatcherConfigValueHolder is just an example. Only ConfigValueHolder matters, as the module manager should be awared of it only.

**Reply by pg-yang** [reply]
> the module manager should be awared of it only.

Yes  . 

And  next question is  how register watchers automatically .  In prepare stage  , we initialize  and  collect  watcher , we should register collected watchers automatically in start stage, but we don't have a global point to execute code

**Reply by wu-sheng** [reply]
We don't. No matter what the object is, the registration process should be taken care of through provider explicitly.

SkyWalking OAP kernel makes sure the boundaries of modules and components clear, in order to make sure we don't use too many automatical mechanism. This is also why we don't try to or directly use stable Spring in the kernel level.

**Reply by pg-yang** [reply]
I  see   , and I would evaluate this function

**Reply by pg-yang** [reply]
> You could have a `ConfigValueHolder` in `library-module`, and when initializing field type, if the type `instanceof`s the `ConfigValueHolder` interface, you use `class#newInstance` and call `ConfigValueHolder#init` to get an instance.

Get an instance of `ConfigValueHolder`?

as you described  ,the field in module config  should  be defined as  
```java
private  SlowDBAccessConfigValueHolder slowDBAccessThreshold;
```
`SlowDBAccessConfigValueHolder` should be defined as 

```java
class SlowDBAccessConfigValueHolder implementation ConfigValueHolder {
    private SlowDBAccessWatcher watcher= null;
    Object init(String propertyValue, Field targetField){
          // new  an instance of watcher , and hold it 
          this.watcher=new SlowDBAccessWatcher(propertyValue);
          // the targetField is defined as  SlowDBAccessConfigValueHolder  
         // so  the return value should be this (or a new SlowDBAccessConfigValueHolder instance  )
          return this;
    }
 // this method used for getting watcher 
   public SlowDBAccessWatcher getWatcher(){
        return this.watcher;
  }

}

```

I read this discussion many times  and still  couldn't  understand it .

**Reply by wu-sheng** [reply]
The above version looks correct already, what else do you not understand?

**Reply by pg-yang** [reply]
This doesn't seem to reduce the code , instead adding an interface . developers need to implement `ConfigValueHolder ` .

**Reply by wu-sheng** [reply]
If your point about simpler is to use fewer codes and make the logic hidden by a powerful but explicit framework, then I am afraid that is a principle differential. As I explained, we are not going to do that. 
At the beginning of this discussion, I was just feeling, you think two fields are differential, then merge them as one. But it seems I misunderstood your original purpose.

**Reply by pg-yang** [reply]
Thanks for your time  , at least  I know the principle
